In [ ]:
!pip install english-words

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 1.1 MB 5.0 MB/s 
  Created wheel for english-words: filename=english_words-1.1.0-py3-none-any.whl size=1106680 sha256=26df0712a930154c17294caab88aec3203c894761fbd2fbbe2807fc6eb4c8e47
  Stored in directory: /root/.cache/pip/wheels/25/3d/4c/12a119ce90b46b4f90f9ddf41d719ecabb40faec6103379fc8
Successfully built english-words


In [ ]:
nltk.download('stopwords')
nltk.download('gutenberg')
nltk.download('wordnet')
nltk.download('omw-1.4')
nltk.download('sentiwordnet')
nltk.download('genesis')
nltk.download('inaugural')
nltk.download('nps_chat')
nltk.download('webtext')
nltk.download('treebank')
from nltk.corpus import wordnet as wn
import nltk
from english_words import english_words_lower_alpha_set
from nltk.wsd import lesk
from nltk.corpus import sentiwordnet as swn
from nltk.corpus import stopwords
from nltk.book import text4
import math

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package sentiwordnet to /root/nltk_data...
[nltk_data]   Package sentiwordnet is already up-to-date!
[nltk_data] Downloading package genesis to /root/nltk_data...
[nltk_data]   Package genesis is already up-to-date!
[nltk_data] Downloading package inaugural to /root/nltk_data...
[nltk_data]   Package inaugural is already up-to-date!
[nltk_data] Downloading package nps_chat to /root/nltk_data...
[nltk_data]   Package nps_chat is already up-to-date!
[nltk_data] Downloading package webtext to /root/nltk_da

#Q1 : Wordnet
Wordnet is a lexical database of the English language, which groups words by part of speech and concept they represent. On the surface level, its similar to a Thesaurus, but Wordnet establishes the semantical relationships between words, while a Thesaurus simply groups words by similar meanings.

#Q2 : Synsets

In [ ]:
noun = "chair"
wn.synsets(noun)

[Synset('chair.n.01'),
 Synset('professorship.n.01'),
 Synset('president.n.04'),
 Synset('electric_chair.n.01'),
 Synset('chair.n.05'),
 Synset('chair.v.01'),
 Synset('moderate.v.01')]

#Q3 : Extract Synset

In [ ]:
print("Definition : " + wn.synset('chair.n.01').definition())
print("Examples : " + str(wn.synset('chair.n.01').examples()))
print("Lemmas : " + str(wn.synset('chair.n.01').lemmas()))
print("Traversing Hierarchy")
curr_set = wn.synset('chair.n.01').hypernyms()
while(len(curr_set) > 0):
  print(curr_set)
  curr_set = curr_set[0].hypernyms()

Definition : a seat for one person, with a support for the back
Examples : ['he put his coat over the back of the chair and sat down']
Lemmas : [Lemma('chair.n.01.chair')]
Traversing Hierarchy
[Synset('seat.n.03')]
[Synset('furniture.n.01')]
[Synset('furnishing.n.02')]
[Synset('instrumentality.n.03')]
[Synset('artifact.n.01')]
[Synset('whole.n.02')]
[Synset('object.n.01')]
[Synset('physical_entity.n.01')]
[Synset('entity.n.01')]


The hierarchy of wordnet is organized into specific objetc, then more and more general terms, until it reaches the point to entity, which is the definition of a noun.

#Q4 : The 'Nyms

In [ ]:
print("Hypernyms : " + str(wn.synset('chair.n.01').hypernyms()))
print("Hyponyms : " + str(wn.synset('chair.n.01').hyponyms()))
print("Meronyms : " + str(wn.synset('chair.n.01').part_meronyms()))
print("Holonyms : " + str(wn.synset('chair.n.01').part_holonyms()))
print("Antonyms : " + str(wn.synset('chair.n.01').lemmas()[0].antonyms()))

Hypernyms : [Synset('seat.n.03')]
Hyponyms : [Synset('armchair.n.01'), Synset('barber_chair.n.01'), Synset('chair_of_state.n.01'), Synset('chaise_longue.n.01'), Synset('eames_chair.n.01'), Synset('fighting_chair.n.01'), Synset('folding_chair.n.01'), Synset('highchair.n.01'), Synset('ladder-back.n.01'), Synset('lawn_chair.n.01'), Synset('rocking_chair.n.01'), Synset('straight_chair.n.01'), Synset('swivel_chair.n.01'), Synset('tablet-armed_chair.n.01'), Synset('wheelchair.n.01')]
Meronyms : [Synset('back.n.08'), Synset('leg.n.03')]
Holonyms : []
Antonyms : []


#Q5 : Verb Synsets

In [ ]:
verb = "run"
verb_synsets = wn.synsets(verb)
print(verb_synsets)

[Synset('run.n.01'), Synset('test.n.05'), Synset('footrace.n.01'), Synset('streak.n.01'), Synset('run.n.05'), Synset('run.n.06'), Synset('run.n.07'), Synset('run.n.08'), Synset('run.n.09'), Synset('run.n.10'), Synset('rivulet.n.01'), Synset('political_campaign.n.01'), Synset('run.n.13'), Synset('discharge.n.06'), Synset('run.n.15'), Synset('run.n.16'), Synset('run.v.01'), Synset('scat.v.01'), Synset('run.v.03'), Synset('operate.v.01'), Synset('run.v.05'), Synset('run.v.06'), Synset('function.v.01'), Synset('range.v.01'), Synset('campaign.v.01'), Synset('play.v.18'), Synset('run.v.11'), Synset('tend.v.01'), Synset('run.v.13'), Synset('run.v.14'), Synset('run.v.15'), Synset('run.v.16'), Synset('prevail.v.03'), Synset('run.v.18'), Synset('run.v.19'), Synset('carry.v.15'), Synset('run.v.21'), Synset('guide.v.05'), Synset('run.v.23'), Synset('run.v.24'), Synset('run.v.25'), Synset('run.v.26'), Synset('run.v.27'), Synset('run.v.28'), Synset('run.v.29'), Synset('run.v.30'), Synset('run.v.31')

#Q6 : Select Verb Synset

In [ ]:
verb_syn = 'run.v.03'
print("Definition : " + wn.synset(verb_syn).definition())
print("Examples : " + str(wn.synset(verb_syn).examples()))
print("Lemmas : " + str(wn.synset(verb_syn).lemmas()))
print("Traversing Hierarchy")
curr_set = wn.synset(verb_syn).hypernyms()
while(len(curr_set) > 0):
  print(str(curr_set[0]) + " = " + str(curr_set[0].definition()))
  curr_set = curr_set[0].hypernyms()

Definition : stretch out over a distance, space, time, or scope; run or extend between two points or beyond a certain point
Examples : ['Service runs all the way to Cranbury', "His knowledge doesn't go very far", 'My memory extends back to my fourth year of life', 'The facts extend beyond a consideration of her personal assets']
Lemmas : [Lemma('run.v.03.run'), Lemma('run.v.03.go'), Lemma('run.v.03.pass'), Lemma('run.v.03.lead'), Lemma('run.v.03.extend')]
Traversing Hierarchy
Synset('be.v.03') = occupy a certain position or area; be somewhere


Unlike Nouns, the verb hierarchy generalizes to "to be", in other words defining verbs in general as a state of being. In this case, "run", or grammatically, "running", is a state of being.

#Q7 : Morphy

In [ ]:
for word in english_words_lower_alpha_set:
  if wn.morphy(word) == "run":
    print(word)

ran
run


#Q8 : Word Similarity

In [ ]:
word_1 = wn.synset('stroll.v.01')
word_2 = wn.synset('walk.v.01')
word_1_leskTest = word_2.definition()
word_2_leskTest = word_1.definition()
print("Wu-Palmer Similarity = " + str(wn.wup_similarity(word_1, word_2)))
lesk_syn_1 = lesk(word_2_leskTest, "walk")
lesk_syn_2 = lesk(word_1_leskTest, "stroll")
print("Lesk Algorithm word_1 = " + str(lesk_syn_1) + " : " + lesk_syn_1.definition())
print("Lesk Algorithm word_2 = " + str(lesk_syn_2) + " : " + lesk_syn_2.definition())

Wu-Palmer Similarity = 0.8
Lesk Algorithm word_1 = Synset('walk.v.10') : take a walk; go for a walk; walk for pleasure
Lesk Algorithm word_2 = Synset('amble.n.01') : a leisurely walk (usually in some public place)


The main difference is that Wu-Palmer compares the similarity of two words, while Lesk attempts to figure out the correct synset of the word based on teh context of the sentence its in. Because of this, in order to test the word similarity, I used the definition of each word as the sentence of the other word, in order to see if they were similar enough that the word in the other sentence would return the synset, but they did not. To be honest, Lesk Algorithm really isnt suited to this.

#Q9 : SentiWordNet
SentiWordNet is basically a lexical analysis method of deducing the emotional charge of a given word, based on the context and the general definitions of said word. One of the best ways to use it is to deduce the tone of a piece of text.

In [ ]:
word = "tantrum"
tantrum = swn.senti_synset('tantrum.n.01')
print(tantrum)
print("Positive score = ", tantrum.pos_score())
print("Negative score = ", tantrum.neg_score())
print("Objective score = ", tantrum.obj_score())
ex_sent = "The boy threw a tantrum because he wanted more candy".split(' ')
for word in ex_sent:
  word1 = swn.senti_synsets(word)
  first = True
  for item in word1:
    print(item)

<fit.n.01: PosScore=0.0 NegScore=0.5>
Positive score =  0.0
Negative score =  0.5
Objective score =  0.5
<male_child.n.01: PosScore=0.25 NegScore=0.0>
<boy.n.02: PosScore=0.0 NegScore=0.0>
<son.n.01: PosScore=0.0 NegScore=0.0>
<boy.n.04: PosScore=0.0 NegScore=0.125>
<throw.v.01: PosScore=0.0 NegScore=0.0>
<throw.v.02: PosScore=0.0 NegScore=0.0>
<shed.v.01: PosScore=0.0 NegScore=0.0>
<throw.v.04: PosScore=0.0 NegScore=0.0>
<give.v.07: PosScore=0.0 NegScore=0.375>
<throw.v.06: PosScore=0.0 NegScore=0.0>
<project.v.10: PosScore=0.0 NegScore=0.0>
<throw.v.08: PosScore=0.0 NegScore=0.0>
<bewilder.v.02: PosScore=0.0 NegScore=0.25>
<hurl.v.03: PosScore=0.125 NegScore=0.0>
<hold.v.03: PosScore=0.125 NegScore=0.0>
<throw.v.12: PosScore=0.0 NegScore=0.0>
<throw.v.13: PosScore=0.0 NegScore=0.0>
<throw.v.14: PosScore=0.0 NegScore=0.0>
<confuse.v.02: PosScore=0.125 NegScore=0.625>
<angstrom.n.01: PosScore=0.0 NegScore=0.0>
<vitamin_a.n.01: PosScore=0.125 NegScore=0.25>
<deoxyadenosine_monophosphate

What the scores tell the program is basically whether each word has a positive or negative connotation, and the intensity of said word. This can allow the program to gain a second layer of comprehension, by allowing it to not just understand literal meaning, but also the meaning hidden in the tone of the sentence.

#Q10 : Collocation
Collocations are phrases, in which the words that make them up have more meaning together than they do apart.

In [ ]:
text4.collocations()

United States; fellow citizens; years ago; four years; Federal
Government; General Government; American people; Vice President; God
bless; Chief Justice; one another; fellow Americans; Old World;
Almighty God; Fellow citizens; Chief Magistrate; every citizen; Indian
tribes; public debt; foreign nations


In [ ]:
text = ' '.join(text4.tokens)
collo = "Chief Justice"
vocab = len(set(text))
hg = text.count('Chief Justice')/vocab
print("p(Chief Justice) = ",hg )
h = text.count('Chief')/vocab
print("p(Chief) = ", h)
g = text.count('Justice')/vocab
print('p(Justice) = ', g)
pmi = math.log2(hg / (h * g))
print('Mutual Information = ', pmi)

p(Chief Justice) =  0.16666666666666666
p(Chief) =  0.3333333333333333
p(Justice) =  0.2857142857142857
Mutual Information =  0.8073549220576041


Based on a Mutual Information that comes to about 0.81, "Chief Justice" is very likely to be a collocation. In an NLTK application, this would be helpful because if the program were to fail to catch collocations, it would lose the comprehension that comes from the added meaning from the combination fo the words.